compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -337,28 +337,17 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
   }
 
   def apiDef(sym: TermSymbol): api.Def = {
-    def paramLists(t: Type, start: Int = 0): List[api.ParameterList] = t match {
+    def paramLists(t: Type, paramss: List[List[Symbol]]): List[api.ParameterList] = t match {
       case pt: TypeLambda =>
-        assert(start == 0)
-        paramLists(pt.resultType)
+        paramLists(pt.resultType, paramss)
       case mt @ MethodTpe(pnames, ptypes, restpe) =>
-        // TODO: We shouldn't have to work so hard to find the default parameters
-        // of a method, Dotty should expose a convenience method for that, see #1143
-        val defaults =
-          if (sym.is(DefaultParameterized)) {
-            val qual =
-              if (sym.isClassConstructor)
-                sym.owner.companionModule // default getters for class constructors are found in the companion object
-              else
-                sym.owner
-            pnames.indices.map(i =>
-              qual.info.member(DefaultGetterName(sym.name, start + i)).exists)
-          } else
-            pnames.indices.map(Function.const(false))
-        val params = pnames.lazyZip(ptypes).lazyZip(defaults).map((pname, ptype, isDefault) =>
-          api.MethodParameter.of(pname.toString, apiType(ptype),
-            isDefault, api.ParameterModifier.Plain))
-        api.ParameterList.of(params.toArray, mt.isImplicitMethod) :: paramLists(restpe, params.length)
+        assert(paramss.nonEmpty && paramss.head.hasSameLengthAs(pnames),
+          i"mismatch for $sym, ${sym.info}, ${sym.paramSymss}")
+        val apiParams = paramss.head.lazyZip(ptypes).map((param, ptype) =>
```

Comment:
@smarter I am not sure whether we need the parameter type `ptype`
in the method type to produce an api type, or whether the type of the
parameter symbol is enough. In the latter case, `paramLists` can be 
much simpler; a simple `nestedMap` instead of the recursion suffices.

Reply:
We need the fully parameter type: if any of its part change, that's a possible API change and some code might need to be recompiled.

Reply:
But `ptype` and the parameter symbol's type are usually isomorphic, right? It's just that one refers to internal parameter symbols and the other to external paramRefs.

Reply:
Ah I see, I haven't checked this PR in detail yet so I can't say but I guess it's fine then

Reply:
I don't know about the fine print though. E.g. does internal vs external representation of varargs matter?

Reply:
As long as the varargs gets represented by an api type which doesn't match some other non-varargs type it should be fine, both representations should have this propriety but I'll double-check. Are there other situations where the types differ?

Reply:
I can't think of anything else.